ouroboros-consensus-cardano/tools/db-analyser/Analysis.hs

```diff
@@ -0,0 +1,232 @@
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module Analysis (HasAnalysis (..)) where
+
+import           Control.Monad.Except
+import qualified Data.Aeson as Aeson
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BL
+import           Data.Foldable (toList)
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           GHC.Natural (Natural)
+
+import           Cardano.Binary (unAnnotated)
+import qualified Cardano.Crypto as Crypto
+
+import qualified Cardano.Chain.Block as Chain
+import qualified Cardano.Chain.Genesis as Genesis
+import qualified Cardano.Chain.Update as Update
+import qualified Cardano.Chain.UTxO as Chain
+
+import qualified Shelley.Spec.Ledger.BlockChain as SL
+import qualified Shelley.Spec.Ledger.PParams as SL
+import qualified Shelley.Spec.Ledger.Tx as SL
+
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Consensus.HardFork.Combinator (OneEraHash (..))
+import           Ouroboros.Consensus.Node.ProtocolInfo
+import           Ouroboros.Consensus.Storage.ChainDB.Serialisation (SizeInBytes)
+
+import           Ouroboros.Consensus.Byron.Ledger (ByronBlock)
+import qualified Ouroboros.Consensus.Byron.Ledger as Byron
+import           Ouroboros.Consensus.Byron.Node (protocolInfoByron)
+
+import           Ouroboros.Consensus.Shelley.Ledger.Block (ShelleyBlock)
+import qualified Ouroboros.Consensus.Shelley.Ledger.Block as Shelley
+import           Ouroboros.Consensus.Shelley.Node (Nonce (..), ShelleyGenesis,
+                     protocolInfoShelley)
+import           Ouroboros.Consensus.Shelley.Protocol.Crypto
+
+import           Ouroboros.Consensus.Cardano.Block (CardanoBlock)
+import qualified Ouroboros.Consensus.Cardano.Block as Cardano
+import           Ouroboros.Consensus.Cardano.Node (TriggerHardFork (..),
+                     protocolInfoCardano)
+
+{-------------------------------------------------------------------------------
+  HasAnalysis
+-------------------------------------------------------------------------------}
+
+class GetPrevHash blk => HasAnalysis blk where
+    countTxOutputs   :: blk -> Int
+    blockHeaderSize  :: blk -> SizeInBytes
+    blockTxSizes     :: blk -> [SizeInBytes]
+    knownEBBs        :: proxy blk -> Map (HeaderHash blk) (ChainHash blk)
```

Comment:
Is the proxy necessary in this case?

Reply:
Hmm, maybe technically speaking no, `ChainHash` fixes `blk`. Just makes it a bit easier to specify which particular set of hashes one is interested in.